sql/analyzer/optimization_rules.go

```diff
@@ -149,21 +149,40 @@ func moveJoinConditionsToFilter(ctx *sql.Context, a *Analyzer, n sql.Node, scope
 		return node, transform.SameTree, nil
 	}
 
-	return transform.NodeWithCtx(node, func(transform.Context) bool { return true }, func(c transform.Context) (sql.Node, transform.TreeIdentity, error) {
-		if c.Node != topJoin {
-			return c.Node, transform.SameTree, nil
+	resultNode, resultIdentity, err := transform.Node(node, func(n sql.Node) (sql.Node, transform.TreeIdentity, error) {
```

Comment:
this creates a CROSS_JOIN at the tree root in testing for me. i suspect the filter would be dropped with the new change
```sql
> explain select * from xy a cross join xy b on a.x = 2;
+-------------------------------------+
| plan                                |
+-------------------------------------+
| CrossJoin                           |
|  ├─ Filter                          |
|  │   ├─ (a.x = 2)                   |
|  │   └─ TableAlias(a)               |
|  │       └─ IndexedTableAccess(xy)  |
|  │           ├─ index: [xy.x]       |
|  │           ├─ filters: [{[2, 2]}] |
|  │           └─ columns: [x y]      |
|  └─ Exchange                        |
|      └─ TableAlias(b)               |
|          └─ Table                   |
|              ├─ name: xy            |
|              └─ columns: [x y]      |
+-------------------------------------+
```

Reply:
you could check if `node == topJoin` before doing the tree search for that case

Reply:
you could also `git checkout main -- sql/analyzer/optimization_rules_test.go` in that case also

Reply:
If I add the `node == topJoin` test, we still get the panic from `populateSubgraph`, even though this seems like the correct action. Investigating this in more detail, we might be trying to fix the wrong issue.

Reply:
> you could also `git checkout main -- sql/analyzer/optimization_rules_test.go` in that case also

Sorry, not following yet, what does this do?

Reply:
if topJoin can be the tree root, the changes the unit tests are unnecessary

Reply:
> you could check if node == topJoin before doing the tree search for that case

This causes a break in a different portion of the code. I suspect one of the downstream rules makes certain assumption about the query plan, and the `node == topJoin` approach violates one of those assumption. As discussed offline a bit, I'm keeping the PR as-is and we'll just use this approach of updating the topJoin parent, instead of updating the topJoin itself.

Reply:
My concerns are correctness related, I don't mind how it happens but CROSS_JOIN root nodes can't lose filters.

Reply:
> My concerns are correctness related, I don't mind how it happens but CROSS_JOIN root nodes can't lose filters.

Absolutely makes sense.
Here's the result of the `explain select...` you mentioned earlier, with this fix in place:
```
+-------------------------------------+
| plan                                |
+-------------------------------------+
| CrossJoin                           |
|  ├─ Filter                          |
|  │   ├─ (a.x = 2)                   |
|  │   └─ TableAlias(a)               |
|  │       └─ IndexedTableAccess(xy)  |
|  │           ├─ index: [xy.x]       |
|  │           ├─ filters: [{[2, 2]}] |
|  │           └─ columns: [x y]      |
|  └─ Exchange                        |
|      └─ TableAlias(b)               |
|          └─ Table                   |
|              ├─ name: xy            |
|              └─ columns: [x y]      |
+-------------------------------------+
```

The filter isn't being dropped. I can add this particular join query to our tests as well.